File: express-server/src/services/pistonService.js
import axios from 'axios';

const PISTON_URL = process.env.PISTON_URL || 'http://localhost:2000';

class PistonService {
    constructor() {
        this.baseURL = `${PISTON_URL}/api/v2`;
    }

    /**
     * Get all available runtimes
     */
    async getRuntimes() {
        try {
            const response = await axios.get(`${this.baseURL}/runtimes`);
            return response.data;
        } catch (error) {
            console.error('Error fetching runtimes:', error.message);
            throw new Error('Failed to fetch available runtimes');
        }
    }

    /**
     * Execute code with Piston
     */
    async executeCode({ language, version, code, stdin = '', args = [] }) {
        try {
                const pistonLanguage = this.getPistonLanguage(language);

            const payload = {
                language: pistonLanguage,
                version,
                files: [
                    {
                        name: this.getFileName(language),
                        content: code
                    }
                ],
                stdin,
                args,
                compile_timeout: 3000,
                run_timeout: 3000,
                compile_memory_limit: -1,
                run_memory_limit: -1
            };

            const response = await axios.post(`${this.baseURL}/execute`, payload);
            return this.formatResponse(response.data);
        } catch (error) {
            console.error('Piston execution error:', error.message);

            if (error.response) {
                return {
                    success: false,
                    error: error.response.data.message || 'Execution failed',
                    stdout: '',
                    stderr: error.response.data.message || 'Unknown error',
                    exitCode: 1
                };
            }

            throw new Error('Failed to execute code');
        }
    }

    /**
     * Format Piston response
     */
    formatResponse(data) {
        const { run, compile } = data;

        let output = '';
        let error = '';
        let success = true;

        // Check compile stage first (for compiled languages)
        if (compile) {
            if (compile.stderr) {
                error += `[Compilation Error]\n${compile.stderr}\n`;
                success = false;
            }
            if (compile.stdout) {
                output += `[Compilation Output]\n${compile.stdout}\n`;
            }
        }

        // Check run stage
        if (run) {
            if (run.stdout) {
                output += run.stdout;
            }
            if (run.stderr) {
                error += run.stderr;
                if (run.code !== 0) {
                    success = false;
                }
            }

            return {
                success: run.code === 0 && !error,
                stdout: output.trim(),
                stderr: error.trim(),
                exitCode: run.code || 0,
                signal: run.signal || null
            };
        }

        return {
            success,
            stdout: output.trim(),
            stderr: error.trim(),
            exitCode: compile?.code || 0
        };
    }

    /**
     * Get appropriate filename based on language
     */
    getFileName(language) {
        const fileNames = {
            javascript: 'main.js',
            typescript: 'main.ts',
            python: 'main.py',
            java: 'Main.java',
            cpp: 'main.cpp',
            c: 'main.c',
            csharp: 'Main.cs',
            go: 'main.go',
            rust: 'main.rs',
            ruby: 'main.rb',
            php: 'main.php',
            swift: 'main.swift',
            kotlin: 'Main.kt',
            bash: 'main.sh',
        };

        return fileNames[language] || 'main.txt';
    }

    /**
     * Get language version mapping
     */
    getLanguageVersion(language) {
        const versionMap = {
            javascript: '18.15.0',
            typescript: '5.0.3',
            python: '3.10.0',
            java: '15.0.2',
            cpp: '10.2.0',
            c: '10.2.0',
            csharp: '6.12.0',
            go: '1.16.2',
            rust: '1.68.2',
            ruby: '3.0.1',
            php: '8.2.3',
            swift: '5.3.3',
            kotlin: '1.8.20',
            bash: '5.2.0',
        };

        return versionMap[language] || '*';
    }

    getPistonLanguage(language) {
        const languageMap = {
            javascript: 'javascript',      // Frontend uses 'javascript', Piston uses 'node'
            typescript: 'typescript',
            python: 'python',
            java: 'java',
            cpp: 'gcc',              // C++ uses gcc
            c: 'gcc',                // C uses gcc
            csharp: 'dotnet',
            go: 'go',
            rust: 'rust',
            ruby: 'ruby',
            php: 'php',
            bash: 'bash',
        };
        return languageMap[language] || language;
    }
}

export default new PistonService();

File: springboot-server/src/main/java/com/codeforge/server/service/PistonService.java
package com.codeforge.server.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;
import org.springframework.http.MediaType;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class PistonService {

    @Value("${piston.url:http://localhost:2000}")
    private String pistonUrl;

    private final RestClient restClient;

    public PistonService(RestClient.Builder builder, @Value("${piston.url:http://localhost:2000}") String pistonBaseUrl) {
        this.restClient = builder.baseUrl(pistonBaseUrl + "/api/v2").build();
    }

    public Object getRuntimes() {
        return restClient.get()
                .uri("/runtimes")
                .retrieve()
                .body(Object.class);
    }

    public Object executeCode(String language, String version, String code, String stdin, List<String> args) {
        String pistonLanguage = getPistonLanguage(language);
        String fileName = getFileName(language);

        Map<String, Object> payload = new HashMap<>();
        payload.put("language", pistonLanguage);
        payload.put("version", version);
        payload.put("files", Collections.singletonList(Map.of("name", fileName, "content", code)));
        payload.put("stdin", stdin != null ? stdin : "");
        payload.put("args", args != null ? args : Collections.emptyList());
        payload.put("compile_timeout", 3000);
        payload.put("run_timeout", 3000);
        payload.put("compile_memory_limit", -1);
        payload.put("run_memory_limit", -1);

        try {
            Map result = restClient.post()
                    .uri("/execute")
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(payload)
                    .retrieve()
                    .body(Map.class);
            return formatResponse(result);
        } catch (Exception e) {
            e.printStackTrace();
            return Map.of(
                    "success", false,
                    "error", e.getMessage(),
                    "stdout", "",
                    "stderr", e.getMessage(),
                    "exitCode", 1
            );
        }
    }

    private Map<String, Object> formatResponse(Map<String, Object> data) {
        Map<String, Object> run = (Map<String, Object>) data.get("run");
        Map<String, Object> compile = (Map<String, Object>) data.get("compile");

        StringBuilder output = new StringBuilder();
        StringBuilder error = new StringBuilder();
        boolean success = true;

        if (compile != null) {
            String stderr = (String) compile.get("stderr");
            String stdout = (String) compile.get("stdout");
            if (stderr != null && !stderr.isEmpty()) {
                error.append("[Compilation Error]\n").append(stderr).append("\n");
                success = false;
            }
            if (stdout != null && !stdout.isEmpty()) {
                output.append("[Compilation Output]\n").append(stdout).append("\n");
            }
        }

        if (run != null) {
            String stdout = (String) run.get("stdout");
            String stderr = (String) run.get("stderr");
            Integer code = (Integer) run.get("code");

            if (stdout != null) {
                output.append(stdout);
            }
            if (stderr != null && !stderr.isEmpty()) {
                error.append(stderr);
                if (code != null && code != 0) {
                    success = false;
                }
            }
            
            return Map.of(
                    "success", success && (code == null || code == 0) && error.length() == 0,
                    "stdout", output.toString().trim(),
                    "stderr", error.toString().trim(),
                    "exitCode", code != null ? code : 0,
                    "signal", run.get("signal") != null ? run.get("signal") : ""
            );
        }

        return Map.of(
                "success", success,
                "stdout", output.toString().trim(),
                "stderr", error.toString().trim(),
                "exitCode", compile != null && compile.get("code") != null ? compile.get("code") : 0
        );
    }

    public String getLanguageVersion(String language) {
        return switch (language) {
            case "javascript" -> "18.15.0";
            case "typescript" -> "5.0.3";
            case "python" -> "3.10.0";
            case "java" -> "15.0.2";
            case "cpp", "c" -> "10.2.0";
            case "csharp" -> "6.12.0";
            case "go" -> "1.16.2";
            case "rust" -> "1.68.2";
            case "ruby" -> "3.0.1";
            case "php" -> "8.2.3";
            case "swift" -> "5.3.3";
            case "kotlin" -> "1.8.20";
            case "bash" -> "5.2.0";
            default -> "*";
        };
    }

    private String getFileName(String language) {
        return switch (language) {
            case "javascript" -> "main.js";
            case "typescript" -> "main.ts";
            case "python" -> "main.py";
            case "java" -> "Main.java";
            case "cpp" -> "main.cpp";
            case "c" -> "main.c";
            case "csharp" -> "Main.cs";
            case "go" -> "main.go";
            case "rust" -> "main.rs";
            case "ruby" -> "main.rb";
            case "php" -> "main.php";
            case "swift" -> "main.swift";
            case "kotlin" -> "Main.kt";
            case "bash" -> "main.sh";
            default -> "main.txt";
        };
    }

    private String getPistonLanguage(String language) {
        return switch (language) {
            case "javascript" -> "javascript";
            case "typescript" -> "typescript";
            case "python" -> "python";
            case "java" -> "java";
            case "cpp", "c" -> "gcc";
            case "csharp" -> "dotnet";
            case "go" -> "go";
            case "rust" -> "rust";
            case "php" -> "php";
            case "bash" -> "bash";
            case "ruby" -> "ruby";
            default -> language;
        };
    }
}
